mlir/lib/Transforms/lab4/savorina_valeria/SavotinaMulAddPass.cpp

```diff
@@ -0,0 +1,82 @@
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Tools/Plugins/PassPlugin.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace mlir;
+
+namespace {
+class SavotinaMulAddPass
+    : public PassWrapper<SavotinaMulAddPass, OperationPass<ModuleOp>> {
+public:
+  StringRef getArgument() const final { return "SavotinaMulAddPass"; }
+  StringRef getDescription() const final { return "fma pass"; }
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<mlir::math::MathDialect>();
+  }
+
+  void runOnOperation() override {
+    mlir::ModuleOp module = getOperation();
+    mlir::OpBuilder builder(module);
+
+    auto replaceAndEraseOp = [&](mlir::LLVM::FMulOp &mulOp,
+                                 mlir::LLVM::FAddOp &addOp,
+                                 mlir::Value &thirdOperand) -> void {
+      builder.setInsertionPoint(addOp);
+      mlir::Value fmaResult =
+          builder.create<mlir::math::FmaOp>(addOp.getLoc(), mulOp.getOperand(0),
+                                            mulOp.getOperand(1), thirdOperand);
+      addOp.replaceAllUsesWith(fmaResult);
+      addOp.erase();
+      mulOp.erase();
```

Comment:
You cannot remove mulOp every time. You need to check for uses of mulOp. You may see this example https://github.com/NN-complr-tech/llvm-nnsu-2024/pull/152#discussion_r1605668496

Reply:
@aobolensk, I don't remove mulOp every time, I check if the result of multiplication is used somewhere else before the addition operation with **isSingleUse**. I added tests relying on example, you can make sure

Reply:
Missed that point, thanks for clarification. Approved